src/Microsoft.ML.OnnxTransform/OnnxTransform.cs

```diff
@@ -531,13 +532,32 @@ public NamedOnnxValue GetNamedOnnxValue()
     /// </summary>
     public sealed class OnnxScoringEstimator : TrivialEstimator<OnnxTransform>
     {
-        public OnnxScoringEstimator(IHostEnvironment env, string modelFile)
-            : this(env, new OnnxTransform(env, modelFile, new string[] { }, new string[] { }))
+        /// <summary>
+        /// Transform for scoring ONNX models. Input data column names/types must exactly match
+        /// all model input names. All possible output columns are generated, with names/types
+        /// specified by model.
+        /// </summary>
+        /// <param name="env">The environment to use.</param>
+        /// <param name="modelFile">Model file path.</param>
+        /// <param name="gpuDeviceId">Optional GPU device ID to run execution on. Null for CPU.</param>
+        /// <param name="fallbackToCpu">If GPU error, raise exception or fallback to CPU.</param>
+        public OnnxScoringEstimator(IHostEnvironment env, string modelFile, int? gpuDeviceId = null, bool fallbackToCpu = false)
```

Comment:
I think it's better for the user to default `fallbackToCpu` to true. This lets the user set GPU and should it fail, let the run continue (hopefully w/ a warning).

Reply:
If a user specifies a GPU id , then the expectation is to run on that device only, and error out if that's not the case.

Setting it to True will incur more 'silent failures', and therefore slower results in general. 



---
In reply to: [245454861](https://github.com/dotnet/machinelearning/pull/2029#discussion_r245454861) [](ancestors = 245454861)

Reply:
Yes, though I'm always advocating for a better user experience. Some image model can be slow enough on a CPU to be a disincentive for users due to runtime. 

I prefer: (1) try to init GPU, (2) use if available; user can override and say use CPU. This causes the happy path to be fast and lets the tyranny of defaults to fall in the user's favor.

Reply:
The 'always-try-gpu-first' may need some thoughts on the following questions:

1> Which deviceID to run on if  multiple are available? Only idle-devices, or in-use-devices are OK as well? 

   -  if in-use-devices are Ok (because they are only lightly loaded), then what are good thresholds to use (e.g. use only if at least 50% GPU memory is free).   It's possible that loading 1 more process onto that device will cause out-of-memory exceptions.

   - If only-idle devices are OK, then should we check that they are idle 'now', or idle in last n-seconds? It is OK if a device has been idle for only 1 micro-second, or should check that it's been idle for 5 minutes?

2> GPU may not be the fastest execution provider. For tiny models, CPU execution may be way faster -- can we tell by model size which one would benefit from GPU? 

3> Also, would/should the user know that this app will grab his GPU by default? Is it OK to do so without their knowledge?

Leaving this value as-is for now -- can update later if needed.


 

---
In reply to: [245460863](https://github.com/dotnet/machinelearning/pull/2029#discussion_r245460863) [](ancestors = 245460863)